packages/query-core/src/types.ts

```diff
@@ -579,19 +579,19 @@ export interface MutateOptions<
   onSuccess?: (
     data: TData,
     variables: TVariables,
-    context: TContext,
-  ) => Promise<unknown> | unknown
+    context: TContext | undefined,
```

Comment:
I think it was on purpose that `onSuccess` always gets the context passed. The type is inferred from whatever `onMutate` returns. It's just potentially undefined in `onError` and `onSettled` because the `onMutate` function is async and can throw / return a rejected Promise, in which case we have no context.

Reply:
Got it, I have created undo commit for this.